JUnit Runner Tests and Missing Hamcrest Documentation

slu-it opened the pull request:
- extended WebTesterJUnitRunner tests and removed redundant validation code for cases that are already checked by the class plausibility checker
- added missing Hamcrest matcher JavaDoc

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/testit-webtester/webtester2-core/14)

<!-- Reviewable:end -->

slu-it committed:
Added big test for JUnit runner and removed redundant code

slu-it committed:
Added missing Hamcrest matcher documentation

slu-it committed:
Removed redundant integration tests

bitterblue commented:
Reviewed 36 of 39 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/AttributeValueMatcher.java, line 31 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN2OeDer-mZ7L1eu6r:-KLN2OeDer-mZ7L1eu6s:-2002080412) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/AttributeValueMatcher.java#L31)):_

> ``` Java
>     private final String attributeName;
>     /** The expected attribute value. */
>     private final String value;
> ```

I'd rename `value` to `expectedValue`.
I also think, the comments for the private fields don't really add value.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/ButtonLabelMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN2t6O-WIURW4rrgr0:-KLN2t6PA-8Mw9Uv6Q6w:-1711124251) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/ButtonLabelMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected label. */
>     private final String label;
> ```

`expectedLabel` for symmetry with `actualLabel`. Or simply `expected` and `actual`, since the class name adds the necessary context for understanding?

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/NumberOfOptionsMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN38RBm9FVCqbl6tWI:-KLN38RBm9FVCqbl6tWJ:-901054577) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/NumberOfOptionsMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected number of options. */
>     private final int numberOfOptions;
> ```

`expectedNumberOfOptions` or simply `expected`?

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/NumberOfSelectedOptionsMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN3DTrcWBMXlPpGlUx:-KLN3DTrcWBMXlPpGlUy:744151562) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/NumberOfSelectedOptionsMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected number of selected options. */
>     private final int numberOfOptions;
> ```

`expectedNumberOfSelectedOptions` or simply `expected`?

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/OptionsTextsMatcher.java, line 28 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN3cl7D2CSgT8Yf2Z7:-KLN3cl8TG6LmvdCwLzL:773299584) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/OptionsTextsMatcher.java#L28)):_

> ``` Java
> 
>     /** The expected list of texts. */
>     private final List<String> texts;
> ```

`expectedTexts`, or just `expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/OptionsValuesMatcher.java, line 28 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN4il-qQEwypx7xrpM:-KLN4il0e3WM-MGvo-SR:-2085094867) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/OptionsValuesMatcher.java#L28)):_

> ``` Java
> 
>     /** The expected list of values. */
>     private final List<String> values;
> ```

`expectedValues`, or `expected`? I really favor `expected` and `actual`, because it nicely lines up for all the matchers.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionIndexMatcher.java, line 26 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN53lfIFTbFnLfsaJA:-KLN53lgFhBQ07h5ITMN:-1314208258) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionIndexMatcher.java#L26)):_

> ``` Java
> 
>     /** The expected index. */
>     private final Integer index;
> ```

`expectedIndex`, or `expected`.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionIndicesMatcher.java, line 28 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5AQQbpn6HRrNMIDN:-KLN5AQRLvZwEoW4dIB7:1674870763) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionIndicesMatcher.java#L28)):_

> ``` Java
> 
>     /** The expected list of indices. */
>     private final List<Integer> indices;
> ```

Again, `expected`, and without the comment :)

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextMatcher.java, line 26 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5LgH1XLyGQ4DyUsz:-KLN5LgH1XLyGQ4DyUt-:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextMatcher.java#L26)):_

> ``` Java
> 
>     /** The expected text. */
>     private final String text;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java, line 14 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5cApqoxd-HBMMcl2:-KLN5cApqoxd-HBMMcl3:-2034968820) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java#L14)):_

> ``` Java
> 
> /**
>  * This {@link TypeSafeMatcher} checks weather or not a {@link MultiSelect} has selected options with a given list of
> ```

weather --> _whether_

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java, line 28 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5UFSNXnprBre5v4Q:-KLN5UFTIgO8GK0gFiZs:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java#L28)):_

> ``` Java
> 
>     /** The expected list of texts. */
>     private final List<String> texts;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionValueMatcher.java, line 26 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5zCwrn7DaMpHRG_w:-KLN5zCwrn7DaMpHRG_x:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionValueMatcher.java#L26)):_

> ``` Java
> 
>     /** The expected value. */
>     private final String value;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionValuesMatcher.java, line 28 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN64jd9ceg6a4hU3A8:-KLN64jd9ceg6a4hU3A9:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionValuesMatcher.java#L28)):_

> ``` Java
> 
>     /** The expected list of values. */
>     private final List<String> values;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TagMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN6Ke-ebRpMDnwVuz_:-KLN6Ke0STaAX6vr0NtT:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TagMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected tag. */
>     private final String tag;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TextContainingMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN6PdmlublGM7YnhD3:-KLN6PdmlublGM7YnhD4:-582283888) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TextContainingMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected text part. */
>     private final String textPart;
> ```

`expectedSubstring`. Hm, this is the first one, where `expected` and `actual` are not symmetric.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TextMatcher.java, line 24 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN6tDu921L3QNXtYS-:-KLN6tDvzBeAatz6xFT0:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/TextMatcher.java#L24)):_

> ``` Java
> 
>     /** The expected text. */
>     private final String text;
> ```

`expected`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/VisibleTextContainingMatcher.java, line 25 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN6z4nr65wF4Wev6nk:-KLN6z4nr65wF4Wev6nl:1457790018) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/VisibleTextContainingMatcher.java#L25)):_

> ``` Java
> 
>     /** The expected text part. */
>     private final String textPart;
> ```

`expectedSubstring`

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/VisibleTextMatcher.java, line 25 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN74a2qxan3OdsjP83:-KLN74a2qxan3OdsjP84:1307922189) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/VisibleTextMatcher.java#L25)):_

> ``` Java
> 
>     /** The expected visible text. */
>     private final String text;
> ```

`expected`

---

_[webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/runner/internal/TestClassPlausibilityChecker.java, line 112 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN9bBlxmDIxM9BDehv:-KLN9bBmGNjb3zZtpmap:-2054912534) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/8f6e645e06bae786927f1c6988b3473ec4a4c353/webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/runner/internal/TestClassPlausibilityChecker.java#L112)):_

> ``` Java
> 
>     private void assertInjectableTypeForConfigurationValueField(Field field) {
>         if (!ConfigurationValueInjector.canInjectValue(field)) {
> ```

I think, this assertion method belongs to `ConfigurationValueInjector`, for two reasons:
- the exception it throws is something, `ConfigurationValueInjector` should throw (the exception message makes it quite clear)
- moving it there would follow "tell, don't ask", which is more robust, because then the key set of the `injectorMap` is completely encapsulated

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14)_

<!-- Sent from Reviewable.io -->

slu-it commented:
Ok, i'll replace fields with actual / expected for the matchers.
And I'll take a look at the TestClassPlausibilityChecker.

There are also some more commits on the way.. should be the last few before the big 2.0.0 release :)

slu-it committed:
Renamed MappingValidatorImpl to DefaultMappingValidator to fit better in the naming scheme for single implementation interfaces

slu-it committed:
Renamed Waiter to DefaultWaiter and extracted interface to fit better in the naming scheme for single implementation interfaces

slu-it committed:
Added tests for the identifications package

- ByProducers factory
- all ByProducer implementations

slu-it committed:
Refactored Mouse API to follow similar principles as the Wait API

- makes it much more flexible
- makes it much more testable
- DefaultMouseDriver is integration tested because of used Selenium classes

slu-it committed:
Fixed typos and removed no longer needed internal utility class TestItHome

slu-it commented:
Review status: 7 of 92 files reviewed at latest revision, 19 unresolved discussions.

---

_[webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java, line 14 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN5cApqoxd-HBMMcl2:-KLNVbyjd9xup34TnsVC:-535708415) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/b26cde8afe89d45b4f9cfa0f1242a40307d4391a/webtester-support-hamcrest/src/main/java/info/novatec/testit/webtester/support/hamcrest/matchers/pagefragments/SelectionTextsMatcher.java#L14)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> weather --> _whether_
> </details>
> that typo existed 29 times in the codebase --- i think i'll never learn ^^

---

_[webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/runner/internal/TestClassPlausibilityChecker.java, line 112 [r2]](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14#-KLN9bBlxmDIxM9BDehv:-KLNVs5J5HaG4sMx2-oc:-1062821824) ([raw file](https://github.com/testit-webtester/webtester2-core/blob/8f6e645e06bae786927f1c6988b3473ec4a4c353/webtester-support-junit4/src/main/java/info/novatec/testit/webtester/junit/runner/internal/TestClassPlausibilityChecker.java#L112)):_
<details><summary>

<i>Previously, bitterblue (Axel Schüssler) wrote…</i></summary>



> I think, this assertion method belongs to `ConfigurationValueInjector`, for two reasons:
> - the exception it throws is something, `ConfigurationValueInjector` should throw (the exception message makes it quite clear)
> - moving it there would follow "tell, don't ask", which is more robust, because then the key set of the `injectorMap` is completely encapsulated
>   </details>
>   moved it to the ConfigurationValueInjector

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14)_

<!-- Sent from Reviewable.io -->

bitterblue commented:
Keep 'em coming :muscle: :grin:

slu-it commented:
I think that should be all of the changes for 2.0.0 - all in all i'm quite content with the current state :)

slu-it commented on webtester-core/src/main/java/info/novatec/testit/webtester/mouse/SequenceWithFragment.java:
```diff
@@ -0,0 +1,139 @@
+ * @since 2.0
+ */
+@Getter(AccessLevel.PACKAGE)
+public class SequenceWithFragment {
```
@bitterblue theoretically this class could extend Sequence. Which would allow us to remove 4 methods since they are 1:1 copies of the corresponding Sequence implementations. What do you think?

bitterblue replied:
Hm, might work out in this case, I think.
In general, switching between return types of fluent API methods communicates context. Depending on context, different methods are valid to be called next (and offered by the IDE's auto-completion). Using inheritance here makes `SequenceWithFragment` a `Sequence`, and it will be impossible to define `Sequence` methods that should never be called in a `SequenceWithFragment` context. So we have less duplicated code at the cost of a potential loss of expressivity. Fluent APIs are all about expressivity, so this feels like a weird trade-off to make... :pensive:

bitterblue replied:
Making `SequenceWithFragment` a `Sequence` means baking in the assumption, that the first in a sequence of mouse actions always involves a page fragment.

slu-it replied:
Hm thats a good point .. haven't thought about that consequence. Reverting on that later on is no option - people might be already using in in breakable ways.
I'll change it back.

slu-it replied:
done

slu-it committed:
Renamed getBrowser() method to browser() since all other similar methods are declared without the "get" prefix as well

slu-it commented:
Ok .. wasn't all after all ... one more and then release!

Drakojin commented:
Reviewed 6 of 39 files at r1, 2 of 3 files at r2, 31 of 31 files at r3, 11 of 13 files at r4, 45 of 47 files at r5, 1 of 2 files at r6, 1 of 1 files at r7, 14 of 14 files at r8.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14)_

<!-- Sent from Reviewable.io -->

bitterblue commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/testit-webtester/webtester2-core/14)_

<!-- Sent from Reviewable.io -->

bitterblue commented:
+1, merge it :)

slu-it commented:
just a little rebase and it's done
